File: apps/android/app/src/test/java/ai/openclaw/app/ui/design/MascotAnimatorTest.kt
package ai.openclaw.app.ui.design

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class MascotAnimatorTest {
  @Test
  fun allMoodChannelsStayInsideBoundsForThirtySeconds() {
    MascotMood.entries.forEachIndexed { index, mood ->
      val animator = MascotAnimator(seed = (index + 1).toULong())
      animator.setMood(mood, 0.0)

      repeat(3_001) { frame ->
        val pose = animator.poseAt(frame / 100.0)
        assertInBounds(pose, mood, frame)
      }
    }
  }

  @Test
  fun sameSeedProducesIdenticalPoses() {
    val first = MascotAnimator(seed = 0xCAFE_BABEuL)
    val second = MascotAnimator(seed = 0xCAFE_BABEuL)
    first.setMood(MascotMood.Curious, 0.0)
    second.setMood(MascotMood.Curious, 0.0)

    repeat(1_500) { frame ->
      val time = frame * 0.023
      if (frame == 400) {
        first.setMood(MascotMood.Thinking, time)
        second.setMood(MascotMood.Thinking, time)
      }
      if (frame == 900) {
        first.setMood(MascotMood.Happy, time)
        second.setMood(MascotMood.Happy, time)
      }
      assertEquals(first.poseAt(time), second.poseAt(time))
    }
  }

  @Test
  fun workingCycleSeatsHatSwingsClawAndShowsWorkEffects() {
    val animator = MascotAnimator(seed = 7uL)
    animator.setMood(MascotMood.Working, 0.0)
    var minRightClaw = Double.POSITIVE_INFINITY
    var maxRightClaw = Double.NEGATIVE_INFINITY
    var seatedHat = false
    var sawSparks = false
    var sawSweat = false

    repeat(2_001) { frame ->
      val time = frame / 100.0
      val pose = animator.poseAt(time)
      minRightClaw = minOf(minRightClaw, pose.rightClawDegrees)
      maxRightClaw = maxOf(maxRightClaw, pose.rightClawDegrees)
      seatedHat = seatedHat || (time >= 1.0 && pose.hardHat >= 0.99)
      sawSparks = sawSparks || pose.effect == MascotEffect.Sparks
      sawSweat = sawSweat || pose.effect == MascotEffect.Sweat
    }

    assertTrue("hard hat never seated", seatedHat)
    assertTrue("hammer swing was ${maxRightClaw - minRightClaw}°", maxRightClaw - minRightClaw > 25.0)
    assertTrue("impact sparks never appeared", sawSparks)
    assertTrue("wipe-brow sweat never appeared", sawSweat)
  }

  @Test
  fun moodChangeCancelsQueuedAndActiveGestures() {
    val animator = MascotAnimator(seed = 11uL)
    animator.poseAt(0.0)
    animator.poseAt(0.95)
    animator.setMood(MascotMood.Thinking, 0.95)

    val afterWaveCancellation = animator.poseAt(1.0)
    assertEquals(0.0, afterWaveCancellation.rightClawDegrees, 0.000_001)

    animator.setMood(MascotMood.Working, 1.1)
    assertTrue(animator.poseAt(1.3).hardHat < 1.0)
    animator.setMood(MascotMood.Sad, 1.3)

    val afterHatCancellation = animator.poseAt(1.31)
    assertEquals(0.0, afterHatCancellation.hardHat, 0.000_001)
  }

  @Test
  fun staticPoseSignaturesMatchMoodContract() {
    assertEquals(MascotPose(), staticPose(MascotMood.Idle))
    assertEquals(MascotGaze(x = 0.3, y = -0.5), staticPose(MascotMood.Thinking).gaze)

    val working = staticPose(MascotMood.Working)
    assertEquals(1.0, working.hardHat, 0.0)
    assertEquals(-28.0, working.rightClawDegrees, 0.0)

    val celebrating = staticPose(MascotMood.Celebrating)
    assertEquals(0.8, celebrating.happyEyes, 0.0)
    assertEquals(30.0, celebrating.leftClawDegrees, 0.0)
    assertEquals(-30.0, celebrating.rightClawDegrees, 0.0)

    val sad = staticPose(MascotMood.Sad)
    assertEquals(0.75, sad.antennaDroop, 0.0)
    assertEquals(-0.55, sad.mouthCurve, 0.0)

    val sleepy = staticPose(MascotMood.Sleepy)
    assertEquals(0.25, sleepy.leftEyeOpenness, 0.0)
    assertEquals(0.5, sleepy.eyeGlowAlpha, 0.0)
  }

  @Test
  fun clampCoversEveryBoundedChannel() {
    val pose =
      MascotPose(
        floatOffset = 100.0,
        antennaDegrees = -100.0,
        antennaDroop = 2.0,
        leftClawDegrees = -100.0,
        rightClawDegrees = 100.0,
        eyeGlowAlpha = -1.0,
        glowScale = 9.0,
        leftEyeOpenness = -1.0,
        rightEyeOpenness = 2.0,
        happyEyes = 2.0,
        gaze = MascotGaze(x = -9.0, y = 9.0),
        mouthCurve = -9.0,
        mouthOpen = 9.0,
        mouthRound = -9.0,
        blush = 9.0,
        hardHat = -9.0,
        bodyTilt = 90.0,
        bodyStretch = 9.0,
        effect = MascotEffect.Sweat,
        effectPhase = 0.75,
      ).clamp()

    assertEquals(2.0, pose.floatOffset, 0.0)
    assertEquals(-14.0, pose.antennaDegrees, 0.0)
    assertEquals(1.0, pose.antennaDroop, 0.0)
    assertEquals(-45.0, pose.leftClawDegrees, 0.0)
    assertEquals(45.0, pose.rightClawDegrees, 0.0)
    assertEquals(0.0, pose.eyeGlowAlpha, 0.0)
    assertEquals(1.6, pose.glowScale, 0.0)
    assertEquals(0.0, pose.leftEyeOpenness, 0.0)
    assertEquals(1.0, pose.rightEyeOpenness, 0.0)
    assertEquals(1.0, pose.happyEyes, 0.0)
    assertEquals(MascotGaze(x = -1.2, y = 1.2), pose.gaze)
    assertEquals(-1.0, pose.mouthCurve, 0.0)
    assertEquals(1.0, pose.mouthOpen, 0.0)
    assertEquals(0.0, pose.mouthRound, 0.0)
    assertEquals(1.0, pose.blush, 0.0)
    assertEquals(0.0, pose.hardHat, 0.0)
    assertEquals(8.0, pose.bodyTilt, 0.0)
    assertEquals(1.05, pose.bodyStretch, 0.0)
    assertEquals(MascotEffect.Sweat, pose.effect)
    assertEquals(0.75, pose.effectPhase, 0.0)
  }

  private fun assertInBounds(
    pose: MascotPose,
    mood: MascotMood,
    frame: Int,
  ) {
    val location = "$mood frame $frame"
    assertTrue(location, pose.floatOffset in -12.0..2.0)
    assertTrue(location, pose.antennaDegrees in -14.0..14.0)
    assertTrue(location, pose.antennaDroop in 0.0..1.0)
    assertTrue(location, pose.leftClawDegrees in -45.0..45.0)
    assertTrue(location, pose.rightClawDegrees in -45.0..45.0)
    assertTrue(location, pose.eyeGlowAlpha in 0.0..1.0)
    assertTrue(location, pose.glowScale in 0.5..1.6)
    assertTrue(location, pose.leftEyeOpenness in 0.0..1.0)
    assertTrue(location, pose.rightEyeOpenness in 0.0..1.0)
    assertTrue(location, pose.happyEyes in 0.0..1.0)
    assertTrue(location, pose.gaze.x in -1.2..1.2)
    assertTrue(location, pose.gaze.y in -1.2..1.2)
    assertTrue(location, pose.mouthCurve in -1.0..1.0)
    assertTrue(location, pose.mouthOpen in 0.0..1.0)
    assertTrue(location, pose.mouthRound in 0.0..1.0)
    assertTrue(location, pose.blush in 0.0..1.0)
    assertTrue(location, pose.hardHat in 0.0..1.0)
    assertTrue(location, pose.bodyTilt in -8.0..8.0)
    assertTrue(location, pose.bodyStretch in 0.86..1.05)
    assertTrue(location, pose.effectPhase in 0.0..1.0)
  }
}

class EffectiveMascotMoodTest {
  @Test
  fun `tinted mascots stay ambient regardless of requested mood`() {
    for (mood in MascotMood.entries) {
      assertEquals(MascotMood.Idle, effectiveMascotMood(mood = mood, tinted = true))
      assertEquals(mood, effectiveMascotMood(mood = mood, tinted = false))
    }
  }
}

File: ui/src/components/mascot-animator.test.ts
import { describe, expect, it } from "vitest";
import { MascotAnimator } from "./mascot-animator.ts";
import { staticMascotPose, type MascotMood, type MascotPose } from "./mascot-pose.ts";

const MOODS: MascotMood[] = [
  "idle",
  "curious",
  "thinking",
  "working",
  "happy",
  "celebrating",
  "sad",
  "sleepy",
  "attentive",
];

function expectPoseInBounds(pose: MascotPose): void {
  expect(pose.floatOffset).toBeGreaterThanOrEqual(-12);
  expect(pose.floatOffset).toBeLessThanOrEqual(2);
  expect(pose.antennaDegrees).toBeGreaterThanOrEqual(-14);
  expect(pose.antennaDegrees).toBeLessThanOrEqual(14);
  expect(pose.antennaDroop).toBeGreaterThanOrEqual(0);
  expect(pose.antennaDroop).toBeLessThanOrEqual(1);
  expect(pose.leftClawDegrees).toBeGreaterThanOrEqual(-45);
  expect(pose.leftClawDegrees).toBeLessThanOrEqual(45);
  expect(pose.rightClawDegrees).toBeGreaterThanOrEqual(-45);
  expect(pose.rightClawDegrees).toBeLessThanOrEqual(45);
  expect(pose.eyeGlowOpacity).toBeGreaterThanOrEqual(0);
  expect(pose.eyeGlowOpacity).toBeLessThanOrEqual(1);
  expect(pose.glowScale).toBeGreaterThanOrEqual(0.5);
  expect(pose.glowScale).toBeLessThanOrEqual(1.6);
  expect(pose.leftEyeOpenness).toBeGreaterThanOrEqual(0);
  expect(pose.leftEyeOpenness).toBeLessThanOrEqual(1);
  expect(pose.rightEyeOpenness).toBeGreaterThanOrEqual(0);
  expect(pose.rightEyeOpenness).toBeLessThanOrEqual(1);
  expect(pose.happyEyes).toBeGreaterThanOrEqual(0);
  expect(pose.happyEyes).toBeLessThanOrEqual(1);
  expect(pose.gaze.x).toBeGreaterThanOrEqual(-1.2);
  expect(pose.gaze.x).toBeLessThanOrEqual(1.2);
  expect(pose.gaze.y).toBeGreaterThanOrEqual(-1.2);
  expect(pose.gaze.y).toBeLessThanOrEqual(1.2);
  expect(pose.mouthCurve).toBeGreaterThanOrEqual(-1);
  expect(pose.mouthCurve).toBeLessThanOrEqual(1);
  expect(pose.mouthOpen).toBeGreaterThanOrEqual(0);
  expect(pose.mouthOpen).toBeLessThanOrEqual(1);
  expect(pose.mouthRound).toBeGreaterThanOrEqual(0);
  expect(pose.mouthRound).toBeLessThanOrEqual(1);
  expect(pose.blush).toBeGreaterThanOrEqual(0);
  expect(pose.blush).toBeLessThanOrEqual(1);
  expect(pose.hardHat).toBeGreaterThanOrEqual(0);
  expect(pose.hardHat).toBeLessThanOrEqual(1);
  expect(pose.bodyTilt).toBeGreaterThanOrEqual(-8);
  expect(pose.bodyTilt).toBeLessThanOrEqual(8);
  expect(pose.bodyStretch).toBeGreaterThanOrEqual(0.86);
  expect(pose.bodyStretch).toBeLessThanOrEqual(1.05);
  expect(pose.dizzy).toBeGreaterThanOrEqual(0);
  expect(pose.dizzy).toBeLessThanOrEqual(1);
}

describe("MascotAnimator", () => {
  it("keeps every mood inside drawable channel bounds for 30 seconds", () => {
    for (const [index, mood] of MOODS.entries()) {
      const animator = new MascotAnimator(1_000 + index);
      animator.setMood(mood, 0);
      for (let frame = 0; frame <= 30 * 30; frame += 1) {
        expectPoseInBounds(animator.poseAt(frame / 30));
      }
    }
  });

  it("produces identical pose streams from the same seed", () => {
    const first = new MascotAnimator(0x5eed);
    const second = new MascotAnimator(0x5eed);
    first.setMood("thinking", 0);
    second.setMood("thinking", 0);

    for (let frame = 0; frame <= 20 * 30; frame += 1) {
      const time = frame / 30;
      expect(first.poseAt(time)).toEqual(second.poseAt(time));
    }
  });

  it("runs the working hard-hat, hammer, impact, and brow-wipe cycle", () => {
    const animator = new MascotAnimator(42);
    animator.setMood("working", 0);
    const rightClawAngles: number[] = [];
    const effects = new Set<string>();
    let seatedHat = 0;

    for (let frame = 0; frame <= 30 * 30; frame += 1) {
      const pose = animator.poseAt(frame / 30);
      rightClawAngles.push(pose.rightClawDegrees);
      effects.add(pose.effect);
      if (frame >= 33) {
        seatedHat = Math.max(seatedHat, pose.hardHat);
      }
    }

    expect(seatedHat).toBe(1);
    expect(Math.max(...rightClawAngles) - Math.min(...rightClawAngles)).toBeGreaterThan(25);
    expect(effects).toContain("sparks");
    expect(effects).toContain("sweat");
  });

  it("cancels stale gestures when the mood changes", () => {
    // Begin as idle (queues the hello wave at +0.9s), then switch mood before
    // it fires — mirrors the element lifecycle where firstUpdated precedes the
    // first `updated()` mood application.
    const animator = new MascotAnimator(9);
    animator.poseAt(0);
    animator.setMood("thinking", 0.1);

    for (let frame = 3; frame <= 4 * 30; frame += 1) {
      const pose = animator.poseAt(frame / 30);
      // The wave raises the right claw far past anything thinking's base loop
      // (claw ~0°) or a scheduled claw snap (-8°) can produce.
      expect(pose.rightClawDegrees).toBeGreaterThan(-15);
    }
  });

  it("keeps sleepy z's visible through its yawn entrance", () => {
    const animator = new MascotAnimator(7);
    animator.setMood("sleepy", 0);
    const pose = animator.poseAt(0.8);

    expect(pose.effect).toBe("zzz");
    expect(pose.mouthRound).toBeGreaterThan(0.8);
    expect(pose.leftEyeOpenness).toBeLessThan(0.1);
  });

  it("overlays and clears the composer tease expression", () => {
    const animator = new MascotAnimator(17);
    animator.setMood("idle", 0);
    animator.setTease(true, 0);

    expect(animator.poseAt(0)).toMatchObject({ mouthRound: 0.5, gaze: { x: 0, y: 0.6 } });

    animator.setTease(false, 0.1);
    const cleared = animator.poseAt(0.1);
    expect(cleared.mouthRound).toBe(0);
    expect(cleared.gaze).not.toEqual({ x: 0, y: 0.6 });
  });

  it("plays one bounded catch beat and clears it after 0.8 seconds", () => {
    const animator = new MascotAnimator(23);
    animator.setMood("idle", 0);
    animator.poseAt(0);
    animator.playCatch(0.1);

    let peakHappyEyes = 0;
    for (let frame = 0; frame <= 24; frame += 1) {
      const pose = animator.poseAt(0.1 + frame / 30);
      expectPoseInBounds(pose);
      peakHappyEyes = Math.max(peakHappyEyes, pose.happyEyes);
    }
    expect(peakHappyEyes).toBeGreaterThan(0.8);
    expect(animator.poseAt(0.91).happyEyes).toBe(0);
    expect(animator.poseAt(1.2).happyEyes).toBe(0);
  });
});

describe("staticMascotPose", () => {
  it("preserves the identifying expression for each motionless mood", () => {
    expect(staticMascotPose("thinking").gaze.y).toBeLessThan(0);
    expect(staticMascotPose("working")).toMatchObject({ hardHat: 1, rightClawDegrees: -28 });
    expect(staticMascotPose("happy").mouthCurve).toBeGreaterThan(0.5);
    expect(staticMascotPose("celebrating").leftClawDegrees).toBeGreaterThan(25);
    expect(staticMascotPose("celebrating").rightClawDegrees).toBeLessThan(-25);
    expect(staticMascotPose("sad")).toMatchObject({ antennaDroop: 0.75, mouthCurve: -0.55 });
    expect(staticMascotPose("sleepy").leftEyeOpenness).toBe(0.25);
    expect(staticMascotPose("attentive")).toEqual(staticMascotPose("idle"));
  });
});
